File: backend/services/income_service.py
"""Income service layer."""

import uuid

from sqlalchemy.ext.asyncio import AsyncSession

from core.logging import logger
from models.income import Income
from repositories.income_repository import income_repository
from schemas.income import IncomeCreate, IncomeUpdate
from utils.exceptions import ResourceNotFoundError, DatabaseError


class IncomeService:
    """Service for income operations."""

    async def get_user_income(
        self,
        session: AsyncSession,
        user_id: uuid.UUID,
        skip: int = 0,
        limit: int = 100,
    ) -> list[Income]:
        """Get all income records for a user."""
        return await income_repository.get_by_user_id(session, user_id, skip, limit)

    async def create_income(
        self,
        session: AsyncSession,
        user_id: uuid.UUID,
        data: IncomeCreate,
    ) -> Income:
        """Create a new income record."""
        try:
            item = await income_repository.create(
                session,
                {
                    "user_id": user_id,
                    "source": data.source,
                    "amount": data.amount,
                    "income_date": data.income_date,
                    "notes": data.notes,
                },
            )
            await session.commit()
            return item
        except Exception as exc:
            logger.error("Failed to create income", extra={"user_id": str(user_id)})
            raise DatabaseError("Failed to create income") from exc

    async def update_income(
        self,
        session: AsyncSession,
        income_id: uuid.UUID,
        user_id: uuid.UUID,
        data: IncomeUpdate,
    ) -> Income:
        """Update an existing income record."""
        try:
            item = await income_repository.get_by_id(session, income_id)
            if item is None or item.user_id != user_id:
                raise ResourceNotFoundError("Income not found")

            update_data = data.model_dump(exclude_unset=True)
            item = await income_repository.update(session, item, update_data)
            await session.commit()
            return item
        except ResourceNotFoundError:
            raise
        except Exception as exc:
            logger.error("Failed to update income", extra={"income_id": str(income_id)})
            raise DatabaseError("Failed to update income") from exc

    async def delete_income(
        self,
        session: AsyncSession,
        income_id: uuid.UUID,
        user_id: uuid.UUID,
    ) -> None:
        """Delete an income record."""
        try:
            item = await income_repository.get_by_id(session, income_id)
            if item is None or item.user_id != user_id:
                raise ResourceNotFoundError("Income not found")

            await income_repository.delete(session, item)
            await session.commit()
        except ResourceNotFoundError:
            raise
        except Exception as exc:
            logger.error("Failed to delete income", extra={"income_id": str(income_id)})
            raise DatabaseError("Failed to delete income") from exc


income_service = IncomeService()

File: frontend/src/services/IncomeService.ts
import { api } from './api'

export interface Income {
  id: string
  userId: string
  source: string
  amount: number
  incomeDate: string
  notes: string | null
  createdAt: string
  updatedAt: string
}

export interface IncomeInput {
  source: string
  amount: number
  incomeDate: string
  notes?: string
}

export const IncomeService = {
  async list(token: string | null): Promise<Income[]> {
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {}
    const response = await api.get('/v1/income', config)
    const payload = response.data?.data
    const items = payload?.items ?? payload
    return Array.isArray(items) ? items : []
  },

  async create(data: IncomeInput, token: string | null): Promise<Income> {
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {}
    const response = await api.post('/v1/income', data, config)
    return response.data?.data
  },

  async update(id: string, data: IncomeInput, token: string | null): Promise<Income> {
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {}
    const response = await api.put(`/v1/income/${id}`, data, config)
    return response.data?.data
  },

  async delete(id: string, token: string | null): Promise<void> {
    const config = token ? { headers: { Authorization: `Bearer ${token}` } } : {}
    await api.delete(`/v1/income/${id}`, config)
  },
}
